File: paper/src/main/java/ru/privatenull/pnauth/paper/PaperPlayerDisplay.java
package ru.privatenull.pnauth.paper;

import net.kyori.adventure.bossbar.BossBar;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.title.Title;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
import ru.privatenull.pnauth.display.ActionBarHandle;
import ru.privatenull.pnauth.display.ActionBarOptions;
import ru.privatenull.pnauth.display.BossBarColor;
import ru.privatenull.pnauth.display.BossBarHandle;
import ru.privatenull.pnauth.display.BossBarOptions;
import ru.privatenull.pnauth.display.BossBarOverlay;
import ru.privatenull.pnauth.display.DisplayHandle;
import ru.privatenull.pnauth.display.Easing;
import ru.privatenull.pnauth.display.PlayerDisplay;
import ru.privatenull.pnauth.display.TitleHandle;
import ru.privatenull.pnauth.display.TitleOptions;
import ru.privatenull.pnauth.platform.PlayerResourceKey;

import java.time.Duration;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

/** Stateful Paper/Folia display implementation backed by Adventure audiences. */
public final class PaperPlayerDisplay implements PlayerDisplay, AutoCloseable {
    private final Plugin plugin;
    private final Map<PlayerResourceKey, Action> actionBars = new ConcurrentHashMap<>();
    private final Map<PlayerResourceKey, PlayerTitle> titles = new ConcurrentHashMap<>();
    private final Map<PlayerResourceKey, PlayerBossBar> bossBars = new ConcurrentHashMap<>();

    public PaperPlayerDisplay(Plugin plugin) {
        this.plugin = plugin;
    }

    @Override
    public ActionBarHandle actionBar(UUID playerId, String displayId, ActionBarOptions options) {
        PlayerResourceKey key = new PlayerResourceKey(playerId, displayId);
        return actionBars.compute(key, (ignored, current) -> {
            if (current == null || !current.active()) return new Action(key, options);
            current.apply(options);
            return current;
        });
    }

    @Override
    public TitleHandle title(UUID playerId, String displayId, TitleOptions options) {
        PlayerResourceKey key = new PlayerResourceKey(playerId, displayId);
        return titles.compute(key, (ignored, current) -> {
            if (current == null || !current.active()) return new PlayerTitle(key, options);
            current.apply(options);
            return current;
        });
    }

    @Override
    public BossBarHandle bossBar(UUID playerId, String displayId, BossBarOptions options) {
        PlayerResourceKey key = new PlayerResourceKey(playerId, displayId);
        return bossBars.compute(key, (ignored, current) -> {
            if (current == null || !current.active()) return new PlayerBossBar(key, options);
            current.apply(options);
            return current;
        });
    }

    @Override public Optional<ActionBarHandle> findActionBar(UUID id, String name) { return Optional.ofNullable(actionBars.get(new PlayerResourceKey(id, name))); }
    @Override public Optional<TitleHandle> findTitle(UUID id, String name) { return Optional.ofNullable(titles.get(new PlayerResourceKey(id, name))); }
    @Override public Optional<BossBarHandle> findBossBar(UUID id, String name) { return Optional.ofNullable(bossBars.get(new PlayerResourceKey(id, name))); }
    @Override public boolean removeActionBar(UUID id, String name) { return close(actionBars.remove(new PlayerResourceKey(id, name))); }
    @Override public boolean removeTitle(UUID id, String name) { return close(titles.remove(new PlayerResourceKey(id, name))); }
    @Override public boolean removeBossBar(UUID id, String name) { return close(bossBars.remove(new PlayerResourceKey(id, name))); }

    @Override
    public void clear(UUID playerId) {
        actionBars.entrySet().removeIf(entry -> closeIfPlayer(entry, playerId));
        titles.entrySet().removeIf(entry -> closeIfPlayer(entry, playerId));
        bossBars.entrySet().removeIf(entry -> closeIfPlayer(entry, playerId));
    }

    @Override
    public void close() {
        actionBars.values().forEach(DisplayHandle::close);
        titles.values().forEach(DisplayHandle::close);
        bossBars.values().forEach(DisplayHandle::close);
        actionBars.clear();
        titles.clear();
        bossBars.clear();
    }

    private abstract class Base implements DisplayHandle {
        protected final PlayerResourceKey key;
        protected volatile boolean active = true;
        protected volatile boolean paused;
        private volatile ScheduledTask expiry;

        protected Base(PlayerResourceKey key, Duration lifetime) {
            this.key = key;
            lifetime(lifetime);
        }
        @Override public UUID playerId() { return key.playerId(); }
        @Override public String displayId() { return key.resourceId(); }
        @Override public boolean active() { return active; }
        @Override public boolean paused() { return paused; }
        @Override public void pause() { paused = true; }
        @Override public void resume() { paused = false; }
        @Override
        public void lifetime(Duration lifetime) {
            cancel(expiry);
            if (valid(lifetime).isZero()) return;
            expiry = schedule(lifetime, this::close);
        }

        protected void withPlayer(java.util.function.Consumer<Player> action) {
            Player player = Bukkit.getPlayer(key.playerId());
            if (player == null) return;
            player.getScheduler().run(plugin, ignored -> action.accept(player), null);
        }

        protected ScheduledTask schedule(Duration delay, Runnable action) {
            Player player = Bukkit.getPlayer(key.playerId());
            if (player == null) return null;
            return player.getScheduler().runDelayed(
                    plugin, ignored -> action.run(), null, ticks(delay));
        }

        protected ScheduledTask repeat(Duration initialDelay, Duration interval, Runnable action) {
            Player player = Bukkit.getPlayer(key.playerId());
            if (player == null) return null;
            return player.getScheduler().runAtFixedRate(
                    plugin, ignored -> action.run(), null, ticks(initialDelay), ticks(interval));
        }

        protected void finish() {
            cancel(expiry);
        }
    }

    private final class Action extends Base implements ActionBarHandle {
        private volatile String text;
        private volatile ScheduledTask refresh;

        private Action(PlayerResourceKey key, ActionBarOptions options) { super(key, options.lifetime()); apply(options); }
        private void apply(ActionBarOptions options) {
            text(options.text());
            refreshInterval(options.refreshInterval());
            lifetime(options.lifetime());
        }
        @Override public void text(String text) { this.text = safe(text); sendNow(); }
        @Override public void refreshInterval(Duration interval) {
            cancel(refresh);
            if (!valid(interval).isZero()) refresh = repeat(interval, interval, this::sendNow);
        }
        @Override public void sendNow() { if (active && !paused) withPlayer(player -> player.sendActionBar(Component.text(text))); }
        @Override public void close() {
            if (!active) return;
            active = false;
            actionBars.remove(key, this);
            cancel(refresh);
            finish();
            withPlayer(player -> player.sendActionBar(Component.empty()));
        }
    }

    private final class PlayerTitle extends Base implements TitleHandle {
        private volatile String title = "";
        private volatile String subtitle = "";
        private volatile Duration fadeIn = Duration.ZERO;
        private volatile Duration stay = Duration.ofSeconds(2);
        private volatile Duration fadeOut = Duration.ZERO;
        private volatile ScheduledTask refresh;

        private PlayerTitle(PlayerResourceKey key, TitleOptions options) { super(key, options.lifetime()); apply(options); }
        private void apply(TitleOptions options) {
            title = safe(options.title()); subtitle = safe(options.subtitle());
            timings(options.fadeIn(), options.stay(), options.fadeOut());
            repeatInterval(options.repeatInterval()); lifetime(options.lifetime()); showNow();
        }
        @Override public void title(String value) { title = safe(value); showNow(); }
        @Override public void subtitle(String value) { subtitle = safe(value); showNow(); }
        @Override public void timings(Duration in, Duration visible, Duration out) { fadeIn = valid(in); stay = valid(visible); fadeOut = valid(out); }
        @Override public void repeatInterval(Duration interval) {
            cancel(refresh);
            if (!valid(interval).isZero()) refresh = repeat(interval, interval, this::showNow);
        }
        @Override public void showNow() {
            if (active && !paused) withPlayer(player -> player.showTitle(Title.title(
                    Component.text(title), Component.text(subtitle), Title.Times.times(fadeIn, stay, fadeOut))));
        }
        @Override public void clear() { withPlayer(Player::clearTitle); }
        @Override public void release() {
            if (!active) return;
            active = false;
            titles.remove(key, this);
            cancel(refresh);
            finish();
        }
        @Override public void close() { if (!active) return; release(); clear(); }
    }

    private final class PlayerBossBar extends Base implements BossBarHandle {
        private final BossBar bar;
        private volatile float progress;
        private volatile ScheduledTask animation;

        private PlayerBossBar(PlayerResourceKey key, BossBarOptions options) {
            super(key, options.lifetime());
            progress = clamp(options.progress());
            bar = BossBar.bossBar(Component.text(options.text()), progress,
                    adventureColor(options.color()), adventureOverlay(options.overlay()));
            properties(options.darkenScreen(), options.playBossMusic(), options.createWorldFog());
            lifetime(options.lifetime());
            withPlayer(player -> player.showBossBar(bar));
        }
        private void apply(BossBarOptions options) {
            text(options.text()); progress(options.progress()); color(options.color()); overlay(options.overlay());
            properties(options.darkenScreen(), options.playBossMusic(), options.createWorldFog());
        }
        @Override public void text(String text) { bar.name(Component.text(safe(text))); }
        @Override public void progress(float progress) { this.progress = clamp(progress); bar.progress(this.progress); }
        @Override public void color(BossBarColor color) { bar.color(adventureColor(color)); }
        @Override public void overlay(BossBarOverlay overlay) { bar.overlay(adventureOverlay(overlay)); }
        @Override public void animateProgress(float target, Duration duration, Easing easing) {
            cancel(animation);
            float start = progress;
            float end = clamp(target);
            int steps = Math.max(1, (int) (valid(duration).toMillis() / 50L));
            AtomicInteger step = new AtomicInteger();
            Easing curve = easing == null ? Easing.LINEAR : easing;
            animation = repeat(Duration.ofMillis(50), Duration.ofMillis(50), () -> {
                if (paused) return;
                float ratio = Math.min(1F, step.incrementAndGet() / (float) steps);
                progress(start + (end - start) * curve.apply(ratio));
                if (ratio >= 1F) cancel(animation);
            });
        }
        @Override public void properties(boolean dark, boolean music, boolean fog) {
            java.util.Set<BossBar.Flag> flags = new java.util.HashSet<>();
            if (dark) flags.add(BossBar.Flag.DARKEN_SCREEN);
            if (music) flags.add(BossBar.Flag.PLAY_BOSS_MUSIC);
            if (fog) flags.add(BossBar.Flag.CREATE_WORLD_FOG);
            bar.flags(flags);
        }
        @Override public void pause() { super.pause(); withPlayer(player -> player.hideBossBar(bar)); }
        @Override public void resume() { super.resume(); withPlayer(player -> player.showBossBar(bar)); }
        @Override public void close() {
            if (!active) return;
            active = false;
            bossBars.remove(key, this);
            cancel(animation);
            finish();
            withPlayer(player -> player.hideBossBar(bar));
        }
    }

    private static boolean close(DisplayHandle handle) { if (handle == null) return false; handle.close(); return true; }
    private static boolean closeIfPlayer(Map.Entry<PlayerResourceKey, ? extends DisplayHandle> entry, UUID playerId) {
        if (!entry.getKey().playerId().equals(playerId)) return false;
        entry.getValue().close(); return true;
    }
    private static String safe(String value) { return value == null ? "" : value; }
    private static Duration valid(Duration value) { return value == null || value.isNegative() ? Duration.ZERO : value; }
    private static long ticks(Duration value) { return Math.max(1L, (valid(value).toMillis() + 49L) / 50L); }
    private static void cancel(ScheduledTask task) { if (task != null) task.cancel(); }
    private static float clamp(float value) { return Math.max(0F, Math.min(1F, value)); }
    private static BossBar.Color adventureColor(BossBarColor value) { return BossBar.Color.valueOf((value == null ? BossBarColor.PURPLE : value).name()); }
    private static BossBar.Overlay adventureOverlay(BossBarOverlay value) { return BossBar.Overlay.valueOf((value == null ? BossBarOverlay.PROGRESS : value).name()); }

}

File: paper/src/main/kotlin/ru/privatenull/pnauth/paper/PaperPlayerDisplay.kt
package ru.privatenull.pnauth.paper

import io.papermc.paper.threadedregions.scheduler.ScheduledTask
import net.kyori.adventure.bossbar.BossBar
import net.kyori.adventure.text.Component
import net.kyori.adventure.title.Title
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import ru.privatenull.pnauth.display.ActionBarHandle
import ru.privatenull.pnauth.display.ActionBarOptions
import ru.privatenull.pnauth.display.BossBarColor
import ru.privatenull.pnauth.display.BossBarHandle
import ru.privatenull.pnauth.display.BossBarOptions
import ru.privatenull.pnauth.display.BossBarOverlay
import ru.privatenull.pnauth.display.DisplayHandle
import ru.privatenull.pnauth.display.Easing
import ru.privatenull.pnauth.display.PlayerDisplay
import ru.privatenull.pnauth.display.TitleBuilder
import ru.privatenull.pnauth.display.TitleHandle
import ru.privatenull.pnauth.display.TitleOptions
import ru.privatenull.pnauth.platform.PlayerResourceKey
import ru.privatenull.pnauth.platform.adapter.PlatformDisplayAdapter
import ru.privatenull.pnauth.message.MessageComponents
import ru.privatenull.pnauth.message.MessageFormat
import java.time.Duration
import java.util.HashSet
import java.util.Optional
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/** Stateful Paper/Folia display implementation backed by Adventure audiences. */
class PaperPlayerDisplay(
    private val plugin: Plugin,
    private val messageFormat: MessageFormat = MessageFormat.MINI_MESSAGE
) : PlatformDisplayAdapter, AutoCloseable {

    override fun showTitle(uniqueId: UUID, builder: TitleBuilder) {
        title(
            uniqueId,
            PLATFORM_TITLE_ID,
            TitleOptions(
                builder.title,
                builder.subtitle,
                builder.fadeIn,
                builder.stay,
                builder.fadeOut,
                Duration.ZERO,
                builder.fadeIn.plus(builder.stay).plus(builder.fadeOut)
            )
        )
    }

    override fun clearTitle(uniqueId: UUID) {
        removeTitle(uniqueId, PLATFORM_TITLE_ID)
    }

    private val actionBars: MutableMap<PlayerResourceKey, Action> = ConcurrentHashMap()
    private val titles: MutableMap<PlayerResourceKey, PlayerTitle> = ConcurrentHashMap()
    private val bossBars: MutableMap<PlayerResourceKey, PlayerBossBar> = ConcurrentHashMap()

    override fun actionBar(playerId: UUID, displayId: String, options: ActionBarOptions): ActionBarHandle {
        val key = PlayerResourceKey(playerId, displayId)
        return actionBars.compute(key) { _, current ->
            if (current == null || !current.active()) Action(key, options)
            else {
                current.applyOptions(options)
                current
            }
        }!!
    }

    override fun title(playerId: UUID, displayId: String, options: TitleOptions): TitleHandle {
        val key = PlayerResourceKey(playerId, displayId)
        return titles.compute(key) { _, current ->
            if (current == null || !current.active()) PlayerTitle(key, options)
            else {
                current.applyOptions(options)
                current
            }
        }!!
    }

    override fun bossBar(playerId: UUID, displayId: String, options: BossBarOptions): BossBarHandle {
        val key = PlayerResourceKey(playerId, displayId)
        return bossBars.compute(key) { _, current ->
            if (current == null || !current.active()) PlayerBossBar(key, options)
            else {
                current.applyOptions(options)
                current
            }
        }!!
    }

    override fun findActionBar(playerId: UUID, displayId: String): Optional<ActionBarHandle> =
        Optional.ofNullable(actionBars[PlayerResourceKey(playerId, displayId)])

    override fun findTitle(playerId: UUID, displayId: String): Optional<TitleHandle> =
        Optional.ofNullable(titles[PlayerResourceKey(playerId, displayId)])

    override fun findBossBar(playerId: UUID, displayId: String): Optional<BossBarHandle> =
        Optional.ofNullable(bossBars[PlayerResourceKey(playerId, displayId)])

    override fun removeActionBar(playerId: UUID, displayId: String): Boolean =
        close(actionBars.remove(PlayerResourceKey(playerId, displayId)))

    override fun removeTitle(playerId: UUID, displayId: String): Boolean =
        close(titles.remove(PlayerResourceKey(playerId, displayId)))

    override fun removeBossBar(playerId: UUID, displayId: String): Boolean =
        close(bossBars.remove(PlayerResourceKey(playerId, displayId)))

    override fun clear(playerId: UUID) {
        actionBars.entries.removeIf { closeIfPlayer(it, playerId) }
        titles.entries.removeIf { closeIfPlayer(it, playerId) }
        bossBars.entries.removeIf { closeIfPlayer(it, playerId) }
    }

    override fun close() {
        actionBars.values.forEach { it.close() }
        titles.values.forEach { it.close() }
        bossBars.values.forEach { it.close() }
        actionBars.clear()
        titles.clear()
        bossBars.clear()
    }

    private abstract inner class Base(
        protected val key: PlayerResourceKey,
        lifetime: Duration?
    ) : DisplayHandle {
        @Volatile protected var isCompletedActive: Boolean = true
        @Volatile protected var isCompletedPaused: Boolean = false
        private @Volatile var expiry: ScheduledTask? = null

        init {
            lifetime(lifetime ?: Duration.ZERO)
        }

        override fun playerId(): UUID = key.playerId
        override fun displayId(): String = key.name
        override fun active(): Boolean = isCompletedActive
        override fun paused(): Boolean = isCompletedPaused

        override fun pause() {
            isCompletedPaused = true
        }

        override fun resume() {
            isCompletedPaused = false
        }

        override fun lifetime(lifetime: Duration) {
            cancel(expiry)
            if (valid(lifetime).isZero) return
            expiry = schedule(lifetime) { close() }
        }

        protected fun withPlayer(action: (Player) -> Unit) {
            val player = Bukkit.getPlayer(key.playerId) ?: return
            player.scheduler.run(plugin, { action(player) }, null)
        }

        protected fun schedule(delay: Duration?, action: Runnable): ScheduledTask? {
            val player = Bukkit.getPlayer(key.playerId) ?: return null
            return player.scheduler.runDelayed(
                plugin, { action.run() }, null, ticks(delay)
            )
        }

        protected fun repeat(initialDelay: Duration?, interval: Duration?, action: Runnable): ScheduledTask? {
            val player = Bukkit.getPlayer(key.playerId) ?: return null
            return player.scheduler.runAtFixedRate(
                plugin, { action.run() }, null, ticks(initialDelay), ticks(interval)
            )
        }

        protected fun finish() {
            cancel(expiry)
        }
    }

    private inner class Action(
        key: PlayerResourceKey,
        options: ActionBarOptions
    ) : Base(key, options.lifetime()), ActionBarHandle {
        @Volatile private var actionText: String = ""
        @Volatile private var refresh: ScheduledTask? = null

        init {
            applyOptions(options)
        }

        fun applyOptions(options: ActionBarOptions) {
            text(options.text())
            refreshInterval(options.refreshInterval())
            lifetime(options.lifetime())
        }

        override fun text(text: String) {
            this.actionText = safe(text)
            sendNow()
        }

        override fun refreshInterval(interval: Duration) {
            cancel(refresh)
            if (!valid(interval).isZero) refresh = repeat(interval, interval) { sendNow() }
        }

        override fun sendNow() {
            if (isCompletedActive && !isCompletedPaused) withPlayer { player -> player.sendActionBar(component(actionText)) }
        }

        override fun close() {
            if (!isCompletedActive) return
            isCompletedActive = false
            actionBars.remove(key, this)
            cancel(refresh)
            finish()
            withPlayer { player -> player.sendActionBar(Component.empty()) }
        }
    }

    private inner class PlayerTitle(
        key: PlayerResourceKey,
        options: TitleOptions
    ) : Base(key, options.lifetime()), TitleHandle {
        @Volatile private var titleText = ""
        @Volatile private var subtitleText = ""
        @Volatile private var fadeIn = Duration.ZERO
        @Volatile private var stay = Duration.ofSeconds(2)
        @Volatile private var fadeOut = Duration.ZERO
        @Volatile private var refresh: ScheduledTask? = null

        init {
            applyOptions(options)
        }

        fun applyOptions(options: TitleOptions) {
            titleText = safe(options.title())
            subtitleText = safe(options.subtitle())
            timings(options.fadeIn(), options.stay(), options.fadeOut())
            repeatInterval(options.repeatInterval())
            lifetime(options.lifetime())
            showNow()
        }

        override fun title(title: String) {
            titleText = safe(title)
            showNow()
        }

        override fun subtitle(subtitle: String) {
            subtitleText = safe(subtitle)
            showNow()
        }

        override fun timings(fadeIn: Duration, stay: Duration, fadeOut: Duration) {
            this.fadeIn = valid(fadeIn)
            this.stay = valid(stay)
            this.fadeOut = valid(fadeOut)
        }

        override fun repeatInterval(interval: Duration) {
            cancel(refresh)
            if (!valid(interval).isZero) refresh = repeat(interval, interval) { showNow() }
        }

        override fun showNow() {
            if (isCompletedActive && !isCompletedPaused) {
                withPlayer { player ->
                    player.showTitle(
                        Title.title(
                            component(titleText),
                            component(subtitleText),
                            Title.Times.times(fadeIn, stay, fadeOut)
                        )
                    )
                }
            }
        }

        override fun clear() {
            withPlayer { it.clearTitle() }
        }

        override fun release() {
            if (!isCompletedActive) return
            isCompletedActive = false
            titles.remove(key, this)
            cancel(refresh)
            finish()
        }

        override fun close() {
            if (!isCompletedActive) return
            release()
            clear()
        }
    }

    private inner class PlayerBossBar(
        key: PlayerResourceKey,
        options: BossBarOptions
    ) : Base(key, options.lifetime()), BossBarHandle {
        private val bar: BossBar
        @Volatile private var currentProgress: Float = clamp(options.progress())
        @Volatile private var animation: ScheduledTask? = null

        init {
            bar = BossBar.bossBar(
                component(options.text()),
                currentProgress,
                adventureColor(options.color()),
                adventureOverlay(options.overlay())
            )
            properties(options.darkenScreen(), options.playBossMusic(), options.createWorldFog())
            lifetime(options.lifetime())
            withPlayer { player -> player.showBossBar(bar) }
        }

        fun applyOptions(options: BossBarOptions) {
            text(options.text())
            progress(options.progress())
            color(options.color())
            overlay(options.overlay())
            properties(options.darkenScreen(), options.playBossMusic(), options.createWorldFog())
        }

        override fun text(text: String) {
            bar.name(component(safe(text)))
        }

        override fun progress(progress: Float) {
            currentProgress = clamp(progress)
            bar.progress(currentProgress)
        }

        override fun color(color: BossBarColor) {
            bar.color(adventureColor(color))
        }

        override fun overlay(overlay: BossBarOverlay) {
            bar.overlay(adventureOverlay(overlay))
        }

        override fun animateProgress(target: Float, duration: Duration, easing: Easing) {
            cancel(animation)
            val start = currentProgress
            val end = clamp(target)
            val steps = Math.max(1, (valid(duration).toMillis() / 50L).toInt())
            val step = AtomicInteger()
            val curve = easing
            animation = repeat(Duration.ofMillis(50), Duration.ofMillis(50)) {
                if (isCompletedPaused) return@repeat
                val ratio = Math.min(1f, step.incrementAndGet() / steps.toFloat())
                progress(start + (end - start) * curve.apply(ratio))
                if (ratio >= 1f) cancel(animation)
            }
        }

        override fun properties(darkenScreen: Boolean, playBossMusic: Boolean, createWorldFog: Boolean) {
            val flags: MutableSet<BossBar.Flag> = HashSet()
            if (darkenScreen) flags.add(BossBar.Flag.DARKEN_SCREEN)
            if (playBossMusic) flags.add(BossBar.Flag.PLAY_BOSS_MUSIC)
            if (createWorldFog) flags.add(BossBar.Flag.CREATE_WORLD_FOG)
            bar.flags(flags)
        }

        override fun pause() {
            super.pause()
            withPlayer { player -> player.hideBossBar(bar) }
        }

        override fun resume() {
            super.resume()
            withPlayer { player -> player.showBossBar(bar) }
        }

        override fun close() {
            if (!isCompletedActive) return
            isCompletedActive = false
            bossBars.remove(key, this)
            cancel(animation)
            finish()
            withPlayer { player -> player.hideBossBar(bar) }
        }
    }

    companion object {
        private const val PLATFORM_TITLE_ID = "pnauth:platform-title"

        private fun close(handle: DisplayHandle?): Boolean {
            if (handle == null) return false
            handle.close()
            return true
        }

        private fun closeIfPlayer(entry: Map.Entry<PlayerResourceKey, DisplayHandle>, playerId: UUID): Boolean {
            if (entry.key.playerId != playerId) return false
            entry.value.close()
            return true
        }

        private fun safe(value: String?): String = value ?: ""
        private fun valid(value: Duration?): Duration = if (value == null || value.isNegative) Duration.ZERO else value
        private fun ticks(value: Duration?): Long = Math.max(1L, (valid(value).toMillis() + 49L) / 50L)
        private fun cancel(task: ScheduledTask?) { task?.cancel() }
        private fun clamp(value: Float): Float = Math.max(0f, Math.min(1f, value))
        private fun adventureColor(value: BossBarColor?): BossBar.Color =
            BossBar.Color.valueOf((value ?: BossBarColor.PURPLE).name)
        private fun adventureOverlay(value: BossBarOverlay?): BossBar.Overlay =
            BossBar.Overlay.valueOf((value ?: BossBarOverlay.PROGRESS).name)
    }

    private fun component(value: String?): Component = MessageComponents.deserialize(value, messageFormat)
}
